src/parse/parser.rs

```diff
@@ -0,0 +1,522 @@
+///  Program for testing and playing with the parser
+///
+
+use std::iter::Iterator;
+use super::ast::{Query, DefStmt, ManipulationStmt, CreateStmt, DropStmt, AltStmt, CreateTableStmt,
+ AlterTableStmt, AlterOp, ColumnInfo, SqlType, UseStmt};
+ use super::token::TokenSpan;
+ use super::lex::Lexer;
+ use std::mem::swap;
+ use super::token::Token;
+ use super::Span;
+
+/*+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+#Parser public functions
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*/
+
+
+// the parser needs a Lexer that iterates through the query
+pub struct Parser<'a> {
+    lexiter: Lexer<'a>,
+
+    last: Option<TokenSpan>,
+    // the current token given by the lexer
+    curr: Option<TokenSpan>,
+    // next token
+    peek: Option<TokenSpan>
+}
+
+impl<'a> Parser<'a> {
+
+    /// Constructs a Parser for the given query.
+    pub fn create(query: &'a str) -> Parser<'a> {
+        let l = Lexer::from_query(query);
+        let mut p = Parser { lexiter: l, last: None, curr: None, peek: None };
+        // Sets initial position of lexer and curr/peek
+        p.lexer_next();
+        p.lexer_next();
+        p
+    }
+
+    /// Parses the given query into an AST
+    pub fn parse(&mut self) -> Result<Query, ParseError>{
+        //deletes Whitespaces in the beginning of Query
+        match self.expect_token(&[Token::Whitespace]) {
+            Ok(&Token::Whitespace) => self.lexer_next(),
+            _=>(),
+        }
+        // first token is checked if it's a keyword using expect_keyword()
+        let querytype = self.expect_keyword(&[Keyword::Create,Keyword::Drop, Keyword::Alter])
+        .map_err(|e| match e {
+            ParseError::UnexpectedEoq => ParseError::EmptyQueryError,
+            _ => e,
+        });
+        // returns high-level AST or Error
```

Comment:
querytype is too long to format it better make the keywords seperate
and split on { with indents
`let keywords = &[Keyword::Create, Keyword::Drop, Keyword::Alter]`
`let querytype = self.expect_keyword(keywords).map_err(|e| match e { `
    `ParseError::UnexpectedEoq => ParseError::EmptyQueryError,_ => e,`
`});`